opal/mca/smsc/cma/smsc_cma_module.c

```diff
@@ -58,6 +66,26 @@ mca_smsc_endpoint_t *mca_smsc_cma_get_endpoint(opal_proc_t *peer_proc)
         return NULL;
     }
 
+#if OPAL_CMA_KCMP_AVAIL
+    /* Check if CAP_SYS_PTRACE capability is allowed between these two processes
+     * Calling process_vm_readv/writev requires CAP_SYS_PTRACE. We can use kcmp
+     * to check if these two processes share a kernel resource. Since kcmp
+     * also requires CAP_SYS_PTRACE it is a good proxy for process_vm_readv/writev.
+     */
+    rc = syscall(SYS_kcmp, getpid(), modex->pid, KCMP_VM, 0, 0);
+    if(rc < 0) {
+        opal_output_verbose(MCA_BASE_VERBOSE_ERROR, opal_smsc_base_framework.framework_output,
```

Comment:
Would `MCA_BASE_VERBOSE_ERRORThi` cause the error to be printed on every attempt? The upper layer can probably recover from it so we don't want to flood the output with error messages. Maybe it's more of a warning that's printed if the user enables verbosity?

Reply:
By default, the `MCA_BASE_VERBOSE_ERROR` will not be printed. The user will need to pass `--mca smsc_base_verbose 1` to see the error. This is because the upper level can recover from the error and work correctly. In the case where `CAP_SYS_PTRACE` is not set, `sm` will still work but it will just not use the `cma` component.

**Example without CAP_SYS_PTRACE and smsc_base_verbose unset**
```
shell$ docker run --rm --user 998:995 ompi-nas-bench mpirun -np 2  --mca pml ob1 --mca btl sm,self /opt/hpc/local/nas/bin/lu.S.x


 NAS Parallel Benchmarks 3.4 -- LU Benchmark

 Size:   12x  12x  12  (class S)
 Iterations:   50
 Total number of processes:      2

 Time step    1
 Time step   20
...
```

**Example without CAP_SYS_PTRACE and smsc_base_verbose set to 1**
```
shell$ docker run --rm --user 998:995 ompi-nas-bench mpirun -np 2 --mca smsc_base_verbose 1 --mca pml ob1 --mca btl sm,self /opt/hpc/local/nas/bin/lu.S.x
[c5e4b54fed21:00010] mca_smsc_cma_module_get_endpoint: can not proceed. processes do not have the necessary permissions (i.e., CAP_SYS_PTRACE). PID 10 <-> 11 (rc = -1) (errno: 1: Operation not permitted)
[c5e4b54fed21:00011] mca_smsc_cma_module_get_endpoint: can not proceed. processes do not have the necessary permissions (i.e., CAP_SYS_PTRACE). PID 11 <-> 10 (rc = -1) (errno: 1: Operation not permitted)


 NAS Parallel Benchmarks 3.4 -- LU Benchmark

 Size:   12x  12x  12  (class S)
 Iterations:   50
 Total number of processes:      2

 Time step    1
 Time step   20
...
```

**Example with CAP_SYS_PTRACE and smsc_base_verbose set to 1**
```
shell$ docker run --rm --user 998:995 --cap-add SYS_PTRACE ompi-nas-bench mpirun -np 2 --mca smsc_base_verbose 1 --mca pml ob1 --mca btl sm,self /opt/hpc/local/nas/bin/lu.S.x


 NAS Parallel Benchmarks 3.4 -- LU Benchmark

 Size:   12x  12x  12  (class S)
 Iterations:   50
 Total number of processes:      2

 Time step    1
 Time step   20
...
```

Reply:
Thanks for the explanation, I wasn't sure whether that is how it works :)